development/build/utils.js

```diff
@@ -118,6 +118,83 @@ function logError(error) {
   console.error(error.stack || error);
 }
 
+/**
+ * This function wrapAgainstScuttling() tries to generically wrap given code
+ * with an environment that allows it to still function under a scuttled environment.
+ *
+ * It's only (current) use is for sentry code which runs before scuttling happens but
+ * later on still leans on properties of the global object which at that point are scuttled.
+ *
+ * To accomplish that, we wrap the entire provided code with the good old with-proxy trick,
+ * which helps us capture access attempts like (1) window.fetch/globalThis.fetch and (2) fetch.
+ *
+ * wrapAgainstScuttling() function also accepts a bag of the global object's properties the
+ * code needs in order to properly function, and within our proxy we make sure to
+ * return those whenever the code goes through our proxy asking for them.
+ *
+ * Specifically when the code tries to set properties to the global object,
+ * in addition to the preconfigured properties, we also accept any property
+ * starting with on to support global event handlers settings.
+ *
+ * Also, sentry invokes functions dynamically using Function.prototype's call and apply,
+ * and our proxy messes with their this when that happens, so these two required a tailor-made patch.
+ *
+ * @param content - contents of the js code to wrap
+ * @param bag - bag of global object properties to provide to the wrapped js code
+ * @returns {string} wrapped js code
+ */
+function wrapAgainstScuttling(content, bag = {}) {
+  return `
+{
+  function setupProxy(global) {
+    // bag of properties to allow vetted shim to access,
+    // mapped to their correct this value if needed
+    const bag = ${JSON.stringify(bag)};
+    // setup vetted shim bag of properties
+    for (const prop in bag) {
+      const that = bag[prop];
+      let api = global[prop];
+      if (that) api = api.bind(global[that]);
+      bag[prop] = api;
+    }
+    // setup proxy for the vetted shim to go through
+    const proxy = new Proxy(bag, {
+      get: function get(target, prop) {
+        return Reflect.get(target, prop);
+      },
+      set: function set(target, prop, value) {
+        if (bag.hasOwnProperty(prop) || prop.startsWith('on')) {
+          return bag[prop] = global[prop] = value;
+        }
+      },
+    });
+    // make sure bind() and apply() are applied with
+    // proxy target rather than proxy receiver
+    (function(target, receiver) {
+      'use strict'; // to work with ses lockdown
+      function wrap(obj, prop, target, receiver) {
```

Comment:
I'm trying to figure out why this is needed - is sentry binding functions to window and giving them away?

Reply:
exactly. but because of our proxy, it ends up binding them to the proxy instead of the window which blows things up. that's why i had to make sure that the bind/apply calls use the window instead of the proxy in case that happens.